### PR TITLE
Allow passing null implementation to generated BindService overload using ServiceBinderBase

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Required to address https://github.com/grpc/grpc-dotnet/issues/21. 

cc @jtattermusch @JamesNK

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @JunTaoLuo isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

File: src/csharp/Grpc.IntegrationTesting/Control.cs [hunk 1]
```diff
@@ -152,6 +153,7 @@ public enum ServerType {
     /// used for some language-specific variants
     /// </summary>
     [pbr::OriginalName("OTHER_SERVER")] OtherServer = 3,
+    [pbr::OriginalName("CALLBACK_SERVER")] CallbackServer = 4,
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.IntegrationTesting/Control.cs]: Not sure why these changes and line endings changes were generated when I ran `generate_proto_csharp.sh`, @jtattermusch ?

Comment (jtattermusch) [reply src/csharp/Grpc.IntegrationTesting/Control.cs]: There might have been some additions changes to the `.proto` file and `.cs` hasn't been regenerated since then. (generally this is fine).

**Comment by JunTaoLuo** [comment]
I signed it

**Comment by JunTaoLuo** [comment]
I've fixed up as many test failures as I can, for the ones I can understand. The rest look like they are unrelated?

**Comment by jtattermusch** [comment]
> I've fixed up as many test failures as I can, for the ones I can understand. The rest look like they are unrelated?

Yes, I think they're unrelated (we has some test infrastructure issues in the past few days). I'll take care of checking the test results before merging.

**Comment by jtattermusch** [comment]
I was thinking through this and I'm still not 100% sure what's the right approach.
My concerns:
- I think having both `BindService` and `BindServiceWithoutImplementation` is too much, that's why I recommended replacing the `BindService` overload instead.
- At the same time, there's value in being able to bind directly to  service implementation class. (because it avoids needing to use reflection to get to the methods once service instance is available).

What do you think about an alternative approach:
- keep the `BindService(grpc::ServiceBinderBase serviceBinder, MathBase serviceImpl)` overload but allow serviceImpl to be null.
- the body of BindService needs to be updated so it accepts serviceImpl == null:
```
// we can't use ?? because the generated code needs to be buildable with older C# lang version
serviceBinder.AddMethod(__Method_Div, serviceImpl != null ? serviceImpl.Div : null);
```
- update the doc comments accordingly

Advantages:
- one can still bind to service implementation if an instance is available
- this prevents users from regenerating the code after the next grpc release, because the BindService method signature stays the same
- (mostly theoretical) in general case, the method name from `.proto` file definition and the C# generated serverside method aren't necessarily the same (there can be some C# source specific escaping in conversion from the Method.Name  to the actual C# serverside method name). The codegen can easily reflect that, but this could be more tricky if accessing the method implementation via reflection.

**Comment by JamesNK** [comment]
Sure, that works too.

**Comment by JunTaoLuo** [comment]
@jtattermusch I considered this design as well (in fact I alluded to it [here](https://github.com/aspnet/grpc-dotnet/pull/9#discussion_r250452736)). Though I ran into a few issues and felt the current approach is simpler and cleaner.

Though you don't have to change the method signature of `BindService`, you do need to change the `AddMethod` signature. Since the expression `serviceImpl != null ? serviceImpl.Div : null` gives an error `Type of conditional expression cannot be determined because there is no implicit conversion between '<null>' and 'method group'`. There are ways around this but we'll probably need to modify how ServiceBinderBase's `AddMethod` signature looks like as they will probably lose the ability to differentiate between different ServerMethods (UnaryServerMethod vs ClientStreamingServerMethod) due to potentially being passed null. Though we can probably address these concerns, it felt like a more significant effort than it's worth, but I may be wrong. I could investigate this approach further if you prefer.

I also wasn't sure how back compat and support for binding to instance were prioritized. Originally I didn't want to remove the overload that binds to an instance since I didn't want to break back compat. However, you mentioned that these methods were originally meant to be used by ASP.NET so it's fine to remove if we don't plan on using them. If we were more concerned with these two points, I would prefer having the two overloads `BindService(grpc::ServiceBinderBase serviceBinder, MathBase serviceImpl)` and `BindService(grpc::ServiceBinderBase serviceBinder)`.

**Comment by JamesNK** [comment]
Once the expression changes from "always a method" to "either a method or null" the compiler isn't able to automatically convert the method to the delegate type. It will need to become something like this:

```cs
serviceBinder.AddMethod(__Method_Chat, serviceImpl != null ? new DuplexStreamingServerMethod<ChatMessage, ChatMessage>(serviceImpl.Chat) : null);
```

**Comment by JunTaoLuo** [comment]
Looks like there's enough information regarding the types during codegen, looks like this will work.

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/17561
https://github.com/grpc/grpc/issues/17814
https://github.com/grpc/grpc/issues/17335
